### ISSUE TITLE
Refactor duplication in overlay RPC

### ISSUE THREAD

**Issue by ogenev** [ethereum/trin#752]
There is quite a duplication code when implementing the base RPC endpoints for every network.

For example, we can see that most of the code in [history rpc ](https://github.com/ethereum/trin/blob/master/trin-history/src/jsonrpc.rs) overlaps with [beacon rpc](https://github.com/ethereum/trin/blob/master/trin-beacon/src/jsonrpc.rs). With the implementation of the State network, this duplication will increase.

I think one piece of refactoring that we want here is to make the common methods generics, i.e. accepting generic `network` and `content_key`/`content_value` arguments. Then we can extract those methods in a separate common module in a way that every network crate has access to it.

**Comment by KolbyML** [comment]
Maybe I could look at this since it would accomplish what I want https://github.com/ethereum/trin/issues/753 cleaner. Let me know if you are already working on this or planning (I might poke at it today) but it is alright either way.

**Comment by rakshit087** [comment]
I am new to rust, so I might be completly wrong. As far as I understand, we can reduce the duplication code by making a trait for "new" function in Beacon Network and HistoryNetwork structs. 
By this we will be able to make the generics strictly typed ( structs that implement the traits ).

**Comment by ogenev** [comment]
> I am new to rust, so I might be completly wrong. As far as I understand, we can reduce the duplication code by making a trait for "new" function in Beacon Network and HistoryNetwork structs. By this we will be able to make the generics strictly typed ( structs that implement the traits ).

Yes and then we can refactor the methods to accept a type that implements this trait and refactor the common code by adding new trait implementation for `HistoryRequestHandler` and `BeaconRequestHandler`.
